### PR TITLE
[WIP] Estimate driving distances between stops

### PR DESCRIPTION
**Pull request by invisiblefunnel**
Fill in missing stop_times.shape_dist_traveled. This approach is almost identical to map-matching, whoa!

Feedback is appreciated :)

### CODE REVIEW

File: partridge/cli.py [hunk 1]
```diff
@@ -0,0 +1,136 @@
...
+            .merge(feed.trips, sort=False)
+            .merge(feed.stops, sort=False)
+            .sort_values('stop_sequence')
+            .groupby('trip_id'))
+
+    candidates = {}
+    for shape_id, points in tqdm(shapes, desc='Prune and cache HMM states'):
+        shape_lons = points.shape_pt_lon.values
+        shape_lats = points.shape_pt_lat.values
+        if shape_lons.shape[0] < 2:
```
Comment (kuanb) [inline partridge/cli.py]: Looks like in this situation we are receiving an array that is greater than 1 dimension. In this case, we have data that is not conforming to expected GTFS spec. In such a case should an error be raised or warning logged - something to inform the operator that the coord value set is being tossed?

File: partridge/cli.py [hunk 2]
```diff
@@ -0,0 +1,136 @@
...
+            continue
+        if pd.isnull(shape_lons).any() or pd.isnull(shape_lats).any():
```
Comment (kuanb) [inline partridge/cli.py]: Same applies here - feedback when tossing candidates would help in debugging.

File: partridge/cli.py [hunk 3]
```diff
@@ -0,0 +1,136 @@
...
+        if shape_id not in candidates:
+            continue
+        if pd.isnull(stop_lons).any() or pd.isnull(stop_lats).any():
+            continue
+        if stop_ids not in observations:
+            observations[stop_ids] = ObservationCache(stop_lons, stop_lats)
+        row_alignments[shape_id, stop_ids].append(np.array(obs[ALIGNMENT_COL]))
+
+    if not any(row_alignments):
+        click.echo('No opportunities to generate shape_dist_traveled.')
```
Comment (kuanb) [inline partridge/cli.py]: Also, might be interesting to report back that, say, `8 out 12 candidates can have shape_dist calculated`.

File: partridge/cli.py [hunk 4]
```diff
@@ -0,0 +1,136 @@
...
+        except NoPathError:
+            dists = np.nan
+
+        for index in indexes:
+            distances[index] = dists
+
+    click.echo('Update shape_dist_traveled values')
+    outfeed.stop_times.set_index(ALIGNMENT_COL, inplace=True)
+    outfeed.stop_times['shape_dist_traveled'] = np.int32(distances)
+    outfeed.stop_times.reset_index(inplace=True)
```
Comment (kuanb) [inline partridge/cli.py]: When resetting index, do we also want to drop the current index (as I believe it was only for the temp purpose  of assigning distance values).

pandas API docs: https://pandas.pydata.org/pandas-docs/stable/generated/pandas.DataFrame.reset_index.html#pandas-dataframe-reset-index (see `drop` defaulted to `False`).

Comment (kuanb) [reply partridge/cli.py]: To that point, so long as row order is preserved - is index setting/resetting even necessary?

File: partridge/cli.py [hunk 5]
```diff
@@ -0,0 +1,136 @@
...
+from tqdm import tqdm
+
+from .config import default_config
+from .gtfs import feed as mkfeed
+from .hmm import (
+    CandidateCache,
+    ObservationCache,
+    MeasureCache,
+    NoPathError,
+    _shape_dist_traveled,
```
Comment (kuanb) [inline partridge/cli.py]: Minor: Why is the `_*` named method being exported/imported. I've always followed the pattern of `_*` methods being restricted to a single file.

File: partridge/hmm.py [hunk 6]
```diff
@@ -0,0 +1,263 @@
...
+
+                    curr_prob = P[astop_idx][aseg_idx]
+                    prob = curr_prob + eprob + tprob
+
+                    if (
+                        bseg_idx not in P[bstop_idx] or
+                        prob > P[bstop_idx][bseg_idx]
+                    ):
+                        P[bstop_idx][bseg_idx] = prob
+                        preds[bstop_idx, bseg_idx].append((
```
[the remark this answers is not in the document]
Comment (invisiblefunnel) [reply partridge/hmm.py]: Ah, you are right. Thanks, Matt.

**Comment by invisiblefunnel** [comment]
@cjer is estimating distances between stops something that would be useful for your work? Would appreciate your feedback on this PR if so.

File: partridge/geo.py [hunk 7]
```diff
@@ -0,0 +1,79 @@
...
+def pairwise_great_circle_vec(lons, lats, earth_radius=EARTH_RADIUS_METERS):
+    assert len(lats) > 1
+    assert len(lats) == len(lons)
+    npoints = len(lats)
+    nsegs = npoints - 1
+    lon1 = np.zeros(nsegs)
+    lat1 = np.zeros(nsegs)
+    lon2 = np.zeros(nsegs)
+    lat2 = np.zeros(nsegs)
+    # There must be a better way to do this with numpy?
```
Comment (kuanb) [inline partridge/geo.py]: How about:
```
def pairwise_great_circle_vec(lons, lats, earth_radius=6371009):
    assert len(lats) > 1
    assert len(lats) == len(lons)
    return great_circle_vec(lons[:-1], lats[:-1], lons[1:], lats[1:], earth_radius)
```

Comment (kuanb) [reply partridge/geo.py]: This method is also more performant:

```
for i in range(10000):
    pairwise_great_circle_vec(lons, lats)
```
CPU times: user 670 ms, sys: 0 ns, total: 670 ms
Wall time: 673 ms

vs.
```
for i in range(10000):
    pairwise_great_circle_vec_alt(lons, lats)
```
CPU times: user 280 ms, sys: 0 ns, total: 280 ms
Wall time: 280 ms

Comment (kuanb) [reply partridge/geo.py]: Another option would be to use something like `pairwise` (http://scikit-learn.org/stable/modules/generated/sklearn.metrics.pairwise.pairwise_distances.html). One way to deal with distance issues in that case would be to recast all coordinate values in, say, an equal area meter projection and leverage the euclidean distance which, for the scale of a bus route, would result in comparable figures without risking degradation in accuracy.

**Comment by cjer** [comment]
@invisiblefunnel afaik our GTFS is not missing shape_dist_traveled, so no immediate need for this. But to see if I understand correctly, you're trying to fill out missing distances between stops on a trip, using an HMM. Can you give a TL;DR for the transition and emission probabilities calculation?

File: partridge/hmm.py [hunk 8]
```diff
@@ -0,0 +1,256 @@
...
+SIGMA = 2.25
+BETA = 18.797498
+
+
+class NoPathError(RuntimeError):
+    pass
+
+
+def shape_dist_traveled(stop_lons, stop_lats, shape_lons, shape_lats,
+                        threshold=200, sigma=SIGMA, beta=BETA):
```
Comment (kuanb) [inline partridge/hmm.py]: I am assuming this is a 200 meter threshold, correct? 

If so, then the process appears to be: 
Take a point, buffer by 200m, and get bounds. Check if bounds intersect with the spatial index developed, which is based on the indexed bounds of each line segment from the route.

So a 2 part question (1 related to how the index is being created, which I will add down below) but, here: 

Should not the buffer length be variable, and a function of the nature of the LineStrings generated? That is, dense networks should have greater sensitivity (e.g. smaller threshold values) whereas comparatively sparse networks can accomodate (perhaps even need) larger threshold values.

Acknowledging this could be a stretch goal for mvp feature and first pass may find threshold default 200 to be acceptable.

File: partridge/hmm.py [hunk 9]
```diff
@@ -0,0 +1,256 @@
...
+class CandidateCache(object):
+    def __init__(self, shape_lons, shape_lats):
+        assert shape_lons.shape[0] > 1
+        assert shape_lons.shape == shape_lats.shape
+        self.shape_lons = shape_lons
+        self.shape_lats = shape_lats
+
+    @cached_property
+    def spatial_index(self):
+        return Index((i, s.bounds, None) for i, s in enumerate(self.segments))
```
Comment (kuanb) [inline partridge/hmm.py]: If my understanding about how `self.segments` works - then is is fair to say that we are generating a bounds sindex for each segment of each linestring in a given route? If so, I think there is a performance opportunity being missed, wherein spatial indices could do more in terms of clustering components of the network (e.g. nodes) and clustering subsegments (single vertices) to create a higher level spatial index. 

Doing so would reduce the number of intersections that would need to be performed. I know a lot of agencies publish shapefiles that are way too detailed. These alignments/shapes would generate spatial indices (in this case) that would be equal in length to the size of the associated shape's LineString vertices length, which would risk really slow execution performance.

File: partridge/hmm.py [hunk 10]
```diff
@@ -0,0 +1,256 @@
...
+    def cumulative_dists(self):
+        return np.cumsum(self.segment_dists)
+
+    @cached_property
+    def segment_dists(self):
+        return pairwise_great_circle_vec(self.shape_lons, self.shape_lats)
+
+    @cached_property
+    def segments(self):
+        return list(map(LineString, pairwise(zip(
```
Comment (kuanb) [inline partridge/hmm.py]: Without digging very deeply, this looks like its creating a bunch of LineStrings that are composed of 2 points (so a singe vertices, essentially, between 2 nodes). Is this correct? If so, is there an opportunity to cluster segments to help create a higher level summary set of grouped route line components for spatial indexing?

File: partridge/geo.py (hunk 7 above)
Comment (invisiblefunnel) [reply partridge/geo.py]: ✅ `great_circle_vec(lons[:-1], lats[:-1], lons[1:], lats[1:], earth_radius)`

That looks great! Care to add that commit to the branch? I made you a collaborator.

Comment (invisiblefunnel) [reply partridge/geo.py]: (no rush btw)

File: partridge/hmm.py (hunk 10 above)
Comment (invisiblefunnel) [reply partridge/hmm.py]: > Is this correct? 

Yep.

> If so, is there an opportunity to cluster segments to help create a higher level summary set of grouped route line components for spatial indexing?

Not sure I follow. How would this work?

File: partridge/cli.py [hunk 11]
```diff
@@ -0,0 +1,136 @@
...
+    feed.shapes
+
+    click.echo('Prepare row alignments for writing')
+    feed.stop_times[ALIGNMENT_COL] = feed.stop_times.index
+    outfeed.stop_times[ALIGNMENT_COL] = outfeed.stop_times.index
+
+    click.echo('Sort and group shapes')
+    shapes = (
+        feed.shapes
+            .sort_values('shape_pt_sequence')
```
[the remark this answers is not in the document]
Comment (invisiblefunnel) [reply partridge/cli.py]: I've always wondered that, but never taken the time to benchmark.

[the remark this answers is not in the document]
Comment (invisiblefunnel) [reply partridge/cli.py]: Here's one opinion: https://github.com/pandas-dev/pandas/issues/15947#issuecomment-292789707

File: partridge/hmm.py (hunk 9 above)
Comment (invisiblefunnel) [reply partridge/hmm.py]: > wherein spatial indices could do more in terms of clustering components of the network (e.g. nodes) and clustering subsegments (single vertices) to create a higher level spatial index.

Can you tell me more about this approach? I'm not sure I understand.

> I know a lot of agencies publish shapefiles that are way too detailed.

Earlier versions of this code used Douglas–Peucker to simplify the lines. I'm not opposed to adding an option to simplify the lines to a given precision, assuming the performance gain is meaningful and it doesn't reduce the overall accuracy. Wdyt?

Comment (kuanb) [reply partridge/hmm.py]: I think a line simplification would be a good first step. I'd suggest converting to a meter projection and simplifying to a ~3 meter (10 ft) resolution.

File: partridge/geo.py [hunk 12]
```diff
@@ -21,21 +27,15 @@ def bbox_from_point(lon, lat, meters):
...
-        lon2[i] = lons[b]
-        lat2[i] = lats[b]
-    return great_circle_vec(lon1, lat1, lon2, lat2, earth_radius)
+    for l in [lats, lons]:
+        if len(l) <= 1:
+            raise InsufficientCoordinateLength(
+                'Not enough coordinates to perform pairwise operation.')
+    if not len(lats) == len(lons):
+        raise UnequalCoordinateLengths(
+            'Different counts of latitudes and longitudes')
```
Comment (invisiblefunnel) [inline partridge/geo.py]: I'd prefer a good doc string to custom errors for this function. It's more plumbing than porcelain and I don't expect users to import it directly.

File: partridge/hmm.py (hunk 9 above)
Comment (invisiblefunnel) [reply partridge/hmm.py]: > I'd suggest converting to a meter projection and simplifying to a ~3 meter (10 ft) resolution.

I like that idea.

File: partridge/geo.py (hunk 12 above)
Comment (kuanb) [reply partridge/geo.py]: Sounds good. I reverted the custom err modifications. I was following a pattern of avoiding asserts and raising custom errors instead.

We could always add in the statement along with the assert, though. E.g.:
```
len(lats) == len(lons), 'Different counts of latitudes and longitudes'
```

File: partridge/hmm.py (hunk 9 above)
Comment (kuanb) [reply partridge/hmm.py]: Welp, just played around with this idea for a little bit to see what gains could be made... and it's now slower. I'm probably making some dumb mistake somewhere, but wanted to share out the progress sketching this out, thus far: https://gist.github.com/kuanb/871c6df0b770f88364aff7a8abc230fc

Comment (invisiblefunnel) [reply partridge/hmm.py]: Oh neat! I took a quick look and I think you're more likely to see the fruits of this work in the graph building and search phases than the intersection. So there may still be hope!

Comment (invisiblefunnel) [reply partridge/hmm.py]: Building the spatial index might also be a tad faster.

**Comment by kuanb** [comment]
Curious if I can help move this forward. Are there a list of targets that need to be met for this to be merged? Performance-related? Considering edge cases (writing tests)?

**Comment by invisiblefunnel** [comment]
Thanks for the nudge, @kuanb. I paused on this to push forward some internal work. Hoping to pick it back up next week.

Right now I'm leaning towards publishing this as a separate PyPi package because it adds some heavy dependencies. Alternatively, it could be an optional dependency to partridge. Wdyt?

**Comment by kuanb** [comment]
@invisiblefunnel gotcha. I am getting on a ~5 hour flight tomorrow so was figuring I could knock some stuff out. :) 

I do see your point about this being an option dep. Seems fine to me. The deps aren't that expensive to install, so I also don't feel this will contribute too much bloat to the lib.

As an aside, I am adding an optional dep (graph-tool) to peartree at the moment, too!

**Comment by invisiblefunnel** [comment]
I'm closing this for now with the intention of implementing it in a separate package.
